pkg/telemetrylogs/json_condition_builder.go

```diff
@@ -0,0 +1,455 @@
+)
+
+var (
+	CodeCurrentNodeNil           = errors.MustNewCode("current_node_nil")
+	CodeNextNodeNil              = errors.MustNewCode("next_node_nil")
+	CodeNestedExpressionsEmpty   = errors.MustNewCode("nested_expressions_empty")
+	CodeGroupByPlanEmpty         = errors.MustNewCode("group_by_plan_empty")
+	CodeArrayMapExpressionsEmpty = errors.MustNewCode("array_map_expressions_empty")
+	CodePromotedPlanMissing      = errors.MustNewCode("promoted_plan_missing")
+	CodeArrayNavigationFailed    = errors.MustNewCode("array_navigation_failed")
+)
+
+func (c *conditionBuilder) getTypes(ctx context.Context, path string) ([]telemetrytypes.JSONDataType, error) {
+	keys, _, err := c.metadataStore.GetKeys(ctx, &telemetrytypes.FieldKeySelector{
+		Name:              path,
+		SelectorMatchType: telemetrytypes.FieldSelectorMatchTypeExact,
+		Signal:            telemetrytypes.SignalLogs,
+		Limit:             1,
+	})
+	if err != nil {
+		return nil, err
+	}
+	types := []telemetrytypes.JSONDataType{}
+	for _, key := range keys[path] {
+		if key.JSONDataType != nil {
+			types = append(types, *key.JSONDataType)
+		}
+	}
+	return types, nil
+}
+
+// BuildCondition builds the full WHERE condition for body_json JSON paths
+func (c *conditionBuilder) buildJSONCondition(ctx context.Context, key *telemetrytypes.TelemetryFieldKey,
+	operator qbtypes.FilterOperator, value any, sb *sqlbuilder.SelectBuilder) (string, error) {
+
+	plan, err := PlanJSON(ctx, key, operator, value, c.getTypes)
+	if err != nil {
+		return "", err
+	}
+
+	conditions := []string{}
+	for _, plan := range plan {
+		condition, err := c.emitPlannedCondition(plan, operator, value, sb)
+		if err != nil {
+			return "", err
+		}
+		conditions = append(conditions, condition)
+	}
+	return sb.Or(conditions...), nil
+}
+
+// emitPlannedCondition handles paths with array traversal
+func (c *conditionBuilder) emitPlannedCondition(plan *telemetrytypes.JSONAccessNode, operator qbtypes.FilterOperator, value any, sb *sqlbuilder.SelectBuilder) (string, error) {
+	// Build traversal + terminal recursively per-hop
+	compiled, err := c.recurseArrayHops(plan, operator, value, sb)
+	if err != nil {
+		return "", err
+	}
+
+	// sb.AddWhereClause(sqlbuilder.NewWhereClause().AddWhereExpr(sb.Args, compiled))
+	return compiled, nil
+}
+
+// buildTerminalCondition creates the innermost condition
+func (c *conditionBuilder) buildTerminalCondition(node *telemetrytypes.JSONAccessNode, operator qbtypes.FilterOperator, value any, sb *sqlbuilder.SelectBuilder) (string, error) {
+	// Use the parent's alias + current field name for the full path
+	fieldPath := node.FieldPath()
+
+	if node.TerminalConfig.ElemType.IsArray {
+		// switch operator for array membership checks
+		switch operator {
+		case qbtypes.FilterOperatorContains, qbtypes.FilterOperatorIn:
+			operator = qbtypes.FilterOperatorEqual
+		case qbtypes.FilterOperatorNotContains, qbtypes.FilterOperatorNotIn:
+			operator = qbtypes.FilterOperatorNotEqual
```

Comment:
(not)contains it makes sense, but `in` doesn't => `=` work, as it would result in invalid query as it build something like `x -> x = ?, ...` with args as array.